### PR TITLE
Added shared module for jq template with SBOM generator

### PR DESCRIPTION
**Pull request by LaurentGoderre**
Input

```
{
    name: "erlang",
    version: "25.3.2.6"
    params: {
        os_name: "alpine",
        os_version: "3.17"
    },
    licenses: [
        "Apache-2.0",
        "MPL-2.0"
    ]
} | sbom | tostring
```

Output:

```json
{
    "spdxVersion": "SPDX-2.3",
    "SPDXID": "SPDXRef-DOCUMENT",
    "name": "erlang-sbom",
    "packages": [
        {
            "name": "erlang",
            "versionInfo": "25.3.2.6",
            "SPDXID": "SPDXRef-Package--erlang",
            "externalRefs": [
                {
                    "referenceCategory": "PACKAGE-MANAGER",
                    "referenceType": "purl",
                    "referenceLocator": "pkg:generic/erlang@25.3.2.6?os_name=alpine&os_version=3.18"
                }
            ],
            "licenseDeclared": "Apache-2.0 AND MPL-2.0"
        }
    ]
}
```

### CODE REVIEW

**Comment by whalelines** [comment]
Tests?

File: scripts/jq-template.jq [hunk 1]
```diff
@@ -0,0 +1,30 @@
+#input package 
+# {
+#     name: "packageName",
+#     version: "packageVersion"
+#     license: "packageLicense"
+# }
+#output: object
+def sbom:
+	. as $input |
```
Comment (whalelines) [inline scripts/jq-template.jq:9]: You don't seem to ever use `$input`.

File: scripts/jq-template.jq [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#input package 
+# {
+#     name: "packageName",
+#     version: "packageVersion"
+#     license: "packageLicense"
```
Comment (whalelines) [inline scripts/jq-template.jq:5]: Include `params` in example input?

**Review by whalelines — COMMENTED** [review]
Is it possible to added tests for this?

**Comment by LaurentGoderre** [comment]
@whalelines I am not seeing other tests for scripts. I will go with what @yosifkit recommends

File: scripts/jq-template.jq [hunk 3]
```diff
@@ -0,0 +1,32 @@
+#input package 
+# {
+#     name: "packageName",
+#     version: "packageVersion",
+#     params: {
+#         "foo": "bar"
+#     }
+#     license: "packageLicense"
```
Comment (whalelines) [inline scripts/jq-template.jq:8]: Seems like this should be `licenses` and its value should be an array?

File: scripts/jq-template.jq [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+				versionInfo: .version,
+				SPDXID: ("SPDXRef-Package--" + .name),
+				externalRefs: [
+					{
+						referenceCategory: "PACKAGE-MANAGER",
+						referenceType: "purl",
+						referenceLocator: ("pkg:generic/" + .name + "@" + .version + "?" + (.params | [to_entries[] | .key + "=" + .value] | join("\u0026")))
+					}
+				],
+				licenseDeclared: .licenses | join(" AND "),
```
Comment (whalelines) [inline scripts/jq-template.jq:28]: Is `licenses` required? If not, a conditional should be added to only add the `licenseDeclared` property if it is not null and not an empty array.

**Review by whalelines — COMMENTED** [review]
Sorry, missed a couple things on my first pass.

File: scripts/jq-template.jq (hunk 4 above)
Comment (LaurentGoderre) [reply scripts/jq-template.jq]: Closest I could get is to output "NOASSERTION" when there is none

File: scripts/jq-template.jq [hunk 5]
```diff
@@ -0,0 +1,36 @@
...
+				name: .name,
+				versionInfo: .version,
+				SPDXID: ("SPDXRef-Package--" + .name),
+				externalRefs: [
+					{
+						referenceCategory: "PACKAGE-MANAGER",
+						referenceType: "purl",
+						referenceLocator: ("pkg:generic/" + .name + "@" + .version + "?" + (.params | [to_entries[] | .key + "=" + .value] | join("\u0026")))
+					}
+				],
```
Comment (yosifkit) [inline scripts/jq-template.jq:27]: Should we add `downloadLocation` if it is given? Are there any other fields that would be useful or interesting to fill in?

**Review by yosifkit — COMMENTED** [review]
:thinking: I think I'd rather name it something descriptive for what it is, like `template-helper-functions.jq`. Maybe we'll add a few more useful global helper functions.

I'm still unsure if this is the best place for the logic. Maybe a scanner improvement to scan the Dockerfile and extract this information directly. I guess it'd have to be basically a shell interpreter 🙈, so maybe not easy to implement.

While the Dockerfile is the source of truth for what was installed and from where, this still adds burden to our external maintainers (and will be unlikely to be caught when it isn't updated on a version bump). Most official-images don't use our custom jq templating (even a few we maintain directly don't).

---

One other item we need to consider is where our tools and scripts will break on a `heredoc`. In reviewing something else, I ran into https://github.com/docker-library/official-images/blob/8a5c557a6a3fdbe8a83c09ec533bee7d3284f29b/diff-pr.sh#L222-L223 which might have issues with a Dockerfile `heredoc`. I have to assume that we may have other places where similar assumptions are made.

**Comment by LaurentGoderre** [comment]
@yosifkit heredoc aren't event required. The following works

```
cat '{{
		{
			name: "docker",
			version: .version,
			params: {
				os_name: "alpine",
				os_version: "3.18"
			},
			licenses: [
				"Apache-2.0"
			]
		} | sbom | tostring
	}}' > /usr/local/bin/docker.spdx.json
```

**Pull request merged**